cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java

```diff
@@ -0,0 +1,159 @@
+import co.cask.cdap.template.etl.api.Validator;
+import co.cask.cdap.template.etl.common.StructuredRecordSerializer;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.script.Invocable;
+import javax.script.ScriptEngine;
+import javax.script.ScriptEngineManager;
+import javax.script.ScriptException;
+
+/**
+ * Validator transformation
+ */
+@Plugin(type = "transform")
+@Name("Validator")
+@Description("Validates a record, writing to an error dataset if the record is invalid. " +
+  "Otherwise it passes the record on to the next stage.")
+public class ValidatorTransform extends Transform<StructuredRecord, StructuredRecord> {
+
+  private static final String SCRIPT_DESCRIPTION = "Javascript that must implement a function 'isValid' that " +
+    "takes a JSON object representation of the input record, " +
+    "and returns Map<String, String> Example : {isValid : false, errorCode : 10, errorMsg : \"unidentified record\"} " +
+    "For example: " +
+    "   function isValid(input) { " +
+    "      var isValid = true; " +
+    "      var errMsg = \"\";" +
+    "      var errCode = 0;" +
+    "      var resultMap = new java.util.HashMap();" +
+    "      input = JSON.parse(input);" +
+    "      if (!apacheValidator.isDate(input.date)) { " +
+    "         isValid = false; errMsg = input.date + \"is invalid date\"; errCode = 5;" +
+    "      } else if (!apacheValidator.isUrl(input.url)) { " +
+    "         isValid = false; errMsg = \"invalid url\"; errCode = 7;" +
+    "      } else if (!apacheValidator.isInRange(input.content_length, 0, 1024 * 1024)) {" +
+    "         isValid = false; errMsg = \"content length >1MB\"; errCode = 10;" +
+    "      }" +
+    "      resultMap.put(\"isValid\", isValid.toString()); " +
+    "      resultMap.put(\"errorCode\", errCode); " +
+    "      resultMap.put(\"errorMsg\", errMsg); " +
+    "      return resultMap;" +
+    "   };" +
+    "This Javascript function in example uses apache validation functions";
+
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(StructuredRecord.class, new StructuredRecordSerializer())
+    .create();
+  private static final Logger LOG = LoggerFactory.getLogger(ValidatorTransform.class);
+
+  private ValidatorConfig config;
+  private Metrics metrics;
+  private Invocable invocable;
+
+  // for unit tests, otherwise config is injected by plugin framework.
+  public ValidatorTransform(ValidatorConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws IllegalArgumentException {
+    for (String validatorName : config.validators.split(",")) {
```

Comment:
a little bit more robust if we split allowing whitespace before and after the comma.